### PR TITLE
Datomic cloud support

### PR DESCRIPTION
**Pull request by mtnygard**
Introduce new literals to allow easy use of Datomic Cloud from Vase + Fern.

* Each of the existing Vase literals now have corresponding "cloud" forms. E.g., `vase.datomic/transact` has a `vase.datomic.cloud/transact`, `vase.datomic/query` has a `vase.datomic.cloud/query`.
* No more AOT! This was causing big problems when the Vase uberjar would have AOT-ed Clojure classes (and every other dependency) but we tried to do development with different library versions.
* This PR also resolves some minor version drift in terms of the Datomic Peer library. In particular, there was a change that makes tempids unnecessary (and even harmful in some cases.) That change is reflected in both the production code and tests now.

### CODE REVIEW

File: pedestal.vase.iml [hunk 1]
```diff
@@ -0,0 +1,145 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ddeaguiar) [inline pedestal.vase.iml]: @mtnygard should we exclude this file from version control?

Comment (mtnygard) [reply pedestal.vase.iml]: Yes, we should. Unfortunately, git submodules on another project kind of screwed me up... which also introduced the spurious commit 08da4eb

Comment (mtnygard) [reply pedestal.vase.iml]: I just pushed a new commit that removes the project file.

File: test/com/cognitect/vase/datomic_cloud_test.clj [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            [clojure.string :as str]
+            [com.cognitect.vase.fern :as vf]
+            [datomic.client.api :as d]
+            [io.pedestal.interceptor :as i]
+            [io.pedestal.interceptor.chain :as chain]))
+
+;; To run these tests, be sure your AWS credentials are available via the environment.
+;; Also, set the envvar DATOMIC_SYSTEM_NAME to the correct name of your Cloud system.
+
+(def db-name (or (System/getenv "VASE_TEST_DB_NAME") "vase-test"))
```
[the remark this answers is not in the document]
Comment (mtnygard) [reply test/com/cognitect/vase/datomic_cloud_test.clj]: I don't really know what we should do with this. The unit tests just use an in-memory DB so no cleanup is needed.

But with the integration tests, we have both the chance of a collision (if someone already has a "vase-test" database in their Datomic Cloud system) and leftover junk when done.

What do you think about generating a unique DB name (like with a UUID in it) at the start of tests then deleting it when done?

File: .gitignore [hunk 3]
```diff
@@ -1,4 +1,6 @@
 # Project related
+.cpcache
```
[the remark this answers is not in the document]
Comment (mtnygard) [reply .gitignore]: OK, I removed the petstore/samples/.cpcache directory from master.

File: test/com/cognitect/vase/datomic_cloud_test.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (ddeaguiar) [reply test/com/cognitect/vase/datomic_cloud_test.clj]: I think that's a good idea.

[the remark this answers is not in the document]
Comment (mtnygard) [reply test/com/cognitect/vase/datomic_cloud_test.clj]: I can't get this to work reliably. Clojure.test has too many ways to throw exceptions that would prevent the cleanup from running. We would end up littering the system with many orphaned databases.

I think it's best at this point to treat this as an integration test with fixed requirements. Suggest we merge this PR and tackle the cleanup problem once everyone has more experience with testing & Datomic Cloud.

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddeaguiar) [reply .gitignore]: 👍

File: deps.edn [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+         io.rkn/conformity        {:mvn/version "0.5.1"}
+
+         ;; Pedestal
+         io.pedestal/pedestal.service {:mvn/version "0.5.4"}
+         io.pedestal/pedestal.jetty   {:mvn/version "0.5.4"}
+
+         com.cognitect/fern {:mvn/version "0.1.3"}
+
+         ;; Nice errors
+         expound {:mvn/version "0.3.1"}}
```
[the remark this answers is not in the document]
Comment (ddeaguiar) [reply deps.edn]: @mtnygard you also need to remove it from this file as well as it's usage [here](https://github.com/cognitect-labs/vase/blob/685cc12a7fdb6e27a191a917680445b197ce53a3/src/com/cognitect/vase/api.clj#L94).

File: samples/petstore/.gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+.cpcache
```
[the remark this answers is not in the document]
Comment (ddeaguiar) [reply samples/petstore/.gitignore]: 👍

File: deps.edn (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtnygard) [reply deps.edn]: I just pushed a correction for deps.edn.

The line you referenced in api.clj was already changed in 
a804c5a. I think you're looking at an old commit.

**Comment by MageMasher** [comment]
Hi @ddeaguiar @mtnygard , what is needed to move this forward? Is there anything I can do? Vase with Cloud is an incredibly exciting opportunity and I'd be happy to contribute where I can.

**Comment by ddeaguiar** [comment]
Hi @MageMasher, I'll need to read through the PR again and follow up with @mtnygard but I think we're close. If there are no additional changes required, it could land on master by Friday.

File: src/com/cognitect/vase/fern.clj [hunk 6]
```diff
@@ -85,7 +87,42 @@
...
+           (log/warn :msg "Cannot execute Tx" :reason :no-connection))
+         ctx)})))
+
+(defmethod f/literal 'vase.datomic.cloud/client [_ client-config db-name]
+  (->CloudConnection client-config db-name))
+
+(defmethod f/literal 'vase.datomic.cloud/attributes [_ ])
+(defmethod f/literal 'vase.datomic.cloud/tx         [_ & assertions] (->CloudTx assertions))
+(defmethod f/literal 'vase.datomic.cloud/attributes [_ & attributes] (->CloudTx (literals/schema-tx attributes)))
+(defmethod f/literal 'vase.datomic.cloud/transact   [_ d] (actions/map->TransactAction (assoc (with-name d) :cloud? true)))
```
Comment (ddeaguiar) [inline src/com/cognitect/vase/fern.clj]: Should be `actions/map->CloudTransactAction`

File: src/com/cognitect/vase/fern.clj [hunk 7]
```diff
@@ -85,7 +87,42 @@
...
+(defmethod f/literal 'vase.datomic.cloud/query      [_ d] (actions/map->QueryAction (assoc (with-name d) :cloud? true)))
```
Comment (ddeaguiar) [inline src/com/cognitect/vase/fern.clj]: Should be `actions/map->CloudQueryAction`

**Comment by ddeaguiar** [comment]
Closing this PR so that I can re-open it against a branch instead of a fork.

**Comment by ddeaguiar** [comment]
This PR's changes were merged in with #99.

**Pull request closed without merging**